tests/lwip_sock_tcp/main.c

```diff
@@ -0,0 +1,1192 @@
+                                          .port = _TEST_PORT_REMOTE,
+                                          .netif = SOCK_ADDR_ANY_NETIF };
+
+    assert(-EAFNOSUPPORT == sock_tcp_connect(&_sock, &remote, 0,
+                                             SOCK_FLAGS_REUSE_EP));
+}
+
+/* ECONNREFUSED does not apply for lwIP; netconn_connect does not wait for
+ * connection build-up */
+
+static void test_tcp_connect4__EINVAL_addr(void)
+{
+    static const sock_tcp_ep_t remote = { .family = AF_INET,
+                                          .port = _TEST_PORT_REMOTE,
+                                          .netif = SOCK_ADDR_ANY_NETIF };
+
+    assert(-EINVAL == sock_tcp_connect(&_sock, &remote, 0, SOCK_FLAGS_REUSE_EP));
+}
+
+static void test_tcp_connect4__EINVAL_netif(void)
+{
+    const sock_tcp_ep_t remote = { .addr = { .ipv4_u32 = HTONL(_TEST_ADDR4_REMOTE) },
+                                          .family = AF_INET,
+                                          .port = _TEST_PORT_REMOTE,
+                                          .netif = (_TEST_NETIF + 1) };
+
+    assert(-EINVAL == sock_tcp_connect(&_sock, &remote, 0, SOCK_FLAGS_REUSE_EP));
+}
+
+/* ENETUNREACH not testable in given loopback setup */
+/* ETIMEDOUT doesn't work because lwIP's connect doesn't timeout */
+
+static void test_tcp_connect4__success_without_port(void)
+{
+    const sock_tcp_ep_t remote = { .addr = { .ipv4_u32 = HTONL(_TEST_ADDR4_REMOTE) },
+                                          .family = AF_INET,
+                                          .port = _TEST_PORT_REMOTE,
+                                          .netif = _TEST_NETIF };
+    msg_t msg = { .type = _SERVER_MSG_START };
+    sock_tcp_ep_t ep;
+
+    _server_addr.family = AF_INET;
+    _server_addr.port = _TEST_PORT_REMOTE;
+    _server_addr.netif = SOCK_ADDR_ANY_NETIF;
+
+    msg_send(&msg, _server);    /* start server on _TEST_PORT_REMOTE */
+
+    assert(0 == sock_tcp_connect(&_sock, &remote, 0, SOCK_FLAGS_REUSE_EP));
+    assert(0 == sock_tcp_get_remote(&_sock, &ep));
+    assert(AF_INET == ep.family);
+    assert(HTONL(_TEST_ADDR4_REMOTE) == ep.addr.ipv4_u32);
+    assert(SOCK_ADDR_ANY_NETIF == ep.netif);
+    assert(_TEST_PORT_REMOTE == ep.port);
+}
+static void test_tcp_connect4__success_local_port(void)
+{
+    const sock_tcp_ep_t remote = { .addr = { .ipv4_u32 = HTONL(_TEST_ADDR4_REMOTE) },
+                                          .family = AF_INET,
+                                          .port = _TEST_PORT_REMOTE,
+                                          .netif = SOCK_ADDR_ANY_NETIF };
+    msg_t msg = { .type = _SERVER_MSG_START };
+    static const uint16_t local_port = _TEST_PORT_LOCAL;
+    sock_tcp_ep_t ep;
+
+    _server_addr.family = AF_INET;
+    _server_addr.port = _TEST_PORT_REMOTE;
+    _server_addr.netif = SOCK_ADDR_ANY_NETIF;
+
+    msg_send(&msg, _server);    /* start server on _TEST_PORT_REMOTE */
```

Comment:
The server is stopped in the `tear_down()` function which is called in between tests.